File: naive/632.binary-tree-maximum-node/Solution.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /*
     * @param root: the root of tree
     * @return: the max node
     */
    public TreeNode maxNode(TreeNode root) {
        // write your code here
        if (root == null){
            return root;
        }
        TreeNode left = maxNode(root.left) ;
        TreeNode right = maxNode(root.right);
        
        return max(root, max(left, right));
    }
    
    public TreeNode max(TreeNode left, TreeNode right){
        if(left==null){
            return right;
        }
        if(right==null){
            return left;
        }
        if(left.val>right.val){
            return left;
        }
        return right;
    }
}

File: naive/632.binary-tree-maximum-node/Solution.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param: root: the root of tree
    @return: the max node
    """
    def maxNode(self, root):
        # write your code here
        if root == None:
            return root
        left = self.maxNode(root.left)
        right = self.maxNode(root.right)
        return self.max_node(root, self.max_node(left, right))
    
    def max_node(self, left, right):
        if left == None:
            return right
        if right == None:
            return left
        if left.val>right.val:
            return left
        return right
        
